packages/client/src/pages/Game/towers/Tower.tsx

```diff
@@ -0,0 +1,243 @@
+    // draw tower object
+    if (this.image) {
+      // save the context
+      this.engine.context?.save();
+      this.engine.context?.translate(
+        this.currentPosition.x - this.towerParams.rectCenterX,
+        this.currentPosition.y - this.towerParams.rectCenterY,
+      );
+      this.engine.context?.rotate(this.towerParams.firingAngle);
+      // this.engine.context?.drawImage(this.image, this.currentPosition.x - this.towerParams.width, this.currentPosition.y - this.towerParams.height, this.towerParams.width, this.towerParams.height)
+      this.engine.context?.drawImage(
+        this.image,
+        -this.towerParams.rectCenterX,
+        -this.towerParams.rectCenterY,
+        this.towerParams.width,
+        this.towerParams.height,
+      );
+      this.engine.context?.closePath();
+      // restore the context
+      this.engine.context?.restore();
+    } else {
+      this.engine.context?.translate(
+        this.currentPosition.x - this.towerParams.rectCenterX,
+        this.currentPosition.y - this.towerParams.rectCenterY,
+      );
+      this.engine.context?.rotate(this.towerParams.firingAngle);
+      this.engine.context!.strokeStyle = this.towerParams.strokeStyle;
+      this.engine.context?.strokeRect(
+        0 - this.towerParams.rectCenterX,
+        0 - this.towerParams.rectCenterY,
+        this.towerParams.width,
+        this.towerParams.height,
+      );
+      this.engine.context?.closePath();
+    }
+  }
+
+  public setAttackInterval = () => {
+    this.attackIntervalTimer = null;
+    // initial build fire
+    this.isCanFire = true;
+    // then set attack interval
+    this.attackIntervalTimer = setInterval(() => {
+      this.isCanFire = true;
+    }, this.towerParams.attackRate);
+  };
+
+  public drawTowerRange() {
+    this.engine.context?.beginPath();
+    this.engine.context!.lineWidth = 0.5;
+    this.engine.context?.setLineDash([10, 15]);
+    this.engine.context!.strokeStyle = this.towerParams.strokeStyle;
+    // draw tower range
+    this.engine.context?.arc(
+      this.currentPosition.x - this.towerParams.width / 2,
+      this.currentPosition.y - this.towerParams.height / 2,
+      this.towerParams.attackRange,
+      0,
+      360,
+    );
+    this.engine.context?.stroke();
+    this.engine.context?.closePath();
+  }
+
+  public drawTower() {
+    // draw tower 2d representation
+    this.draw();
+    // draw tower range
+    if (this.engine.isCanBuild) {
+      this.drawTowerRange();
+    }
+  }
+
+  public isEnemyInRange(enemy: Enemy) {
+    const distance =
+      (enemy.currentPosition.x -
+        this.currentPosition.x +
+        this.towerParams.width) *
+        (enemy.currentPosition.x -
+          this.currentPosition.x +
+          this.towerParams.width) +
+      (enemy.currentPosition.y -
+        this.currentPosition.y +
+        this.towerParams.height) *
+        (enemy.currentPosition.y -
+          this.currentPosition.y +
+          this.towerParams.height);
```

Comment:
Maybe here we could use Math.pow ?
```suggestion
      Math.pow((enemy.currentPosition.x -
        this.currentPosition.x +
        this.towerParams.width), 2) +
      Math.pow((enemy.currentPosition.y -
        this.currentPosition.y +
        this.towerParams.height), 2);
```

Reply:
Math.pow is too expensive at CPU time, a*a is faster then Math.pow(a, 2)

Howewer, completely refactored for result of
`const xDistance =
      enemy.currentPosition.x - this.currentPosition.x + this.towerParams.width;`
    `const yDistance =
      enemy.currentPosition.y -
      this.currentPosition.y +
      this.towerParams.height;`
   ` if (Math.hypot(xDistance, yDistance) < this.towerParams.attackRange) {
      this.target = enemy;
      return true;
    }
    return false;`